src/enclave/enclave_oe.c

```diff
@@ -1,27 +1,144 @@
+#include <enclave/oe_compat.h>
+
 #include <string.h>
 
 #include "pthread_impl.h"
 
+#include <openenclave/bits/eeid.h>
 #include <openenclave/internal/globals.h>
 
 #include "enclave/enclave_oe.h"
 #include "enclave/enclave_signal.h"
 #include "enclave/enclave_util.h"
-#include "enclave/sgxlkl_config.h"
 #include "shared/env.h"
 
-extern int sgxlkl_verbose;
+const sgxlkl_enclave_config_t* sgxlkl_enclave = NULL;
+sgxlkl_enclave_state_t sgxlkl_enclave_state = {0};
+
+bool sgxlkl_in_sw_debug_mode()
+{
+    return sgxlkl_enclave_state.config->mode == SW_DEBUG_MODE;
+}
+
+bool sgxlkl_in_hw_debug_mode()
+{
+    return sgxlkl_enclave_state.config->mode == HW_DEBUG_MODE;
+}
+
+bool sgxlkl_in_hw_release_mode()
+{
+    return sgxlkl_enclave_state.config->mode == HW_RELEASE_MODE;
+}
+
+static void prepare_elf_stack()
+{
+    sgxlkl_enclave_state_t* state = &sgxlkl_enclave_state;
+    const sgxlkl_enclave_config_t* cfg = state->config;
+
+    // import host envp
+    state->num_imported_env = 0;
+    state->imported_env = NULL;
+
+    if (sgxlkl_enclave_state.shared_memory.env && cfg->num_host_import_env > 0)
```

Comment:
That happens when the host refuses to share it's `envp`. We have to assume it's malicious, so the more checks the better. (In fact, it could change it's mind and replace it's `envp` while the loop below executes - see comment about deep copy of shared memory in `enclave_oe.c`).

Reply:
I see, that's a tricky one, I'm not sure deep copy is a good idea since you could also flood the enclave with that by creating a large number of env vars on the host.
@prp @davidchisnall Is this fine as it is?

Reply:
I think that we should deep-copy all outside data (and enforce a size limit), otherwise we are inviting TOCTOU attacks etc.

Reply:
Yeah, I'm getting ever closer to thinking I should just deep-copy the entire state even if it's perhaps not strictly required.

Reply:
I'm fine with that but we need to be careful with the env vars because people put all kinds of big stuff in there nowadays. If we have a size limit then I think the launcher should already filter the env vars down to the set of imported ones.

Reply:
Yeah, that would be nice too. We have to filter and copy them in the enclave anyways though as we can't rely on the host being nice and we need to copy those that we want onto the ELF stack.

Reply:
Absolutely, we need to filter/check twice. This would just avoid sending more than needed to the enclave.

Reply:
(Deep copy is done, no energy for pre-filtering on the host left.)